### PR TITLE
add component_is_part_of_tensor

### PR DESCRIPTION
**Pull request by davydden**
fixes https://github.com/dealii/dealii/issues/6800

TODO:

- [x] get it working
- [x] rename `vector_data_range` into `non_scalar`
- [x] use the corresponding enum
- [x] separate reflow of comment into stand-alone PRs
- [x] cleanup the test

### CODE REVIEW

**Comment by jppelteret** [comment]
Is there any chance that you'd be willing to add the equivalent for symmetric tensors while you're busy putting this together? I'm guessing there'd be a large overlap in the logic for these two cases, and you'd be touching the same parts of the code.

**Comment by davydden** [comment]
> Is there any chance that you'd be willing to add the equivalent for symmetric tensors while you're busy putting this together?

Likely not. It already a huge PITA and reminds me the questions of complex-valued numbers two years ago. But it should be easy to add later if/when someone needs this, assuming that I make the whole thing working...

**Comment by davydden** [comment]
finally found some example of how to output tensors in VTK https://www.paraview.org/pipermail/paraview/2017-February/039214.html Will update this post if I find more...

File: include/deal.II/base/data_out_base.h [hunk 1]
```diff
@@ -1587,13 +1587,13 @@ namespace DataOutBase
...
-    const std::vector<Patch<dim, spacedim>> &patches,
-    const std::vector<std::string> &         data_names,
-    const std::vector<std::tuple<unsigned int, unsigned int, std::string>>
-      &            vector_data_ranges,
-    const DXFlags &flags,
-    std::ostream & out);
+  write_dx(const std::vector<Patch<dim, spacedim>> &patches,
+           const std::vector<std::string> &         data_names,
+           const std::vector<
+             std::tuple<unsigned int, unsigned int, std::string, unsigned int>>
```
[the remark this answers is not in the document]
Comment (davydden) [reply include/deal.II/base/data_out_base.h]: > do we document somewhere what the elements of the tuple refer to

yes, we do. it's range of components (inclusive), i.e. `[0,2]` for 3D vec. 

> looks funny

yes, but that's the downside of usage of weak enumerators. I really want to store here the component interpretation. 

If you want, I am more than happy to switch to strong enumerators later on in this PR.

File: source/base/data_out_base.cc [hunk 2]
```diff
@@ -3948,25 +3791,16 @@ namespace DataOutBase
...
-              // by Stefan Nauber. please no-one
-              // ask me why it works that way (or
-              // may be not), especially not about
-              // the angles and the sign of
-              // the height field, I don't know
-              // it.
+              // this computation was first written by Stefan Nauber. please
+              // no-one ask me why it works that way (or may be not), especially
+              // not about the angles and the sign of the height field, I don't
+              // know it.
```
Comment (bangerth) [inline source/base/data_out_base.cc]: here and everywhere else, did you just re-flow all of these comments by hand? The patch is hard to read because the real meat is hidden between all of these unrelated changes. If you had a way to break things out and into a separate patch, that would be swell.

File: include/deal.II/base/data_out_base.h [hunk 3]
```diff
@@ -3104,7 +3113,8 @@ class DataOutReader : public DataOutInterface<dim, spacedim>
    * string, meaning that all data is to be considered a collection of scalar
    * fields.
    */
-  virtual std::vector<std::tuple<unsigned int, unsigned int, std::string>>
+  virtual std::vector<
+    std::tuple<unsigned int, unsigned int, std::string, unsigned int>>
   get_vector_data_ranges() const override;
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/data_out_base.h]: Oh, I see that this is 2eb72e043082faf253f9f1468a7ba3bad2d2b753. Make this into a separate pull request if you can and lets get it out of the way.

File: source/base/data_out_base.cc (hunk 2 above)
Comment (davydden) [reply source/base/data_out_base.cc]: i will pull-out a few things once I get it working...

File: include/deal.II/base/data_out_base.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/data_out_base.h]: Oh, so are you saying that the last element of the tuple is the component_interpretation? If so, why not use the corresponding `enum`?

I think I was confused by the fact that the function (and the arguments' name) is still called `vector_data_range` when it's not just vectors any more. Maybe this should then be called `non_scalar` now?

[the remark this answers is not in the document]
Comment (davydden) [reply include/deal.II/base/data_out_base.h]: > If so, why not use the corresponding enum?

i guess i confused myself by assuming that has to be a strong enum.

> Maybe this should then be called non_scalar now?

yes, they should.

File: source/base/data_out_base.cc (hunk 2 above)
Comment (davydden) [reply source/base/data_out_base.cc]: p.s. 

> re-flow all of these comments by hand?

no, clang-format

Comment (bangerth) [reply source/base/data_out_base.cc]: I must have not paid attention to this -- I thought we already did this for all files in the switch to clang-format. Or did you just trigger this for one file?

Comment (davydden) [reply source/base/data_out_base.cc]: >  thought we already did this for all files in the switch to clang-format.

AFAIK this won't happen unless you are over the limit. So you need to trigger re-flow but creating long lines.

@tamiko might know more.

File: include/deal.II/base/data_out_base.h [hunk 4]
```diff
@@ -2850,20 +2858,21 @@ class DataOutInterface
...
    *
    * Since some of the derived classes do not know about vector data, this
    * function has a default implementation that simply returns an empty
    * string, meaning that all data is to be considered a collection of scalar
    * fields.
    */
-  virtual std::vector<std::tuple<unsigned int, unsigned int, std::string>>
+  virtual std::vector<
+    std::tuple<unsigned int, unsigned int, std::string, unsigned int>>
   get_vector_data_ranges() const;
```
[the remark this answers is not in the document]
Comment (davydden) [reply include/deal.II/base/data_out_base.h]: > I think you ought to document what the last element now is

btw, it's actually **is** documented 😄 But I still will rename it for clarity.

**Comment by jppelteret** [comment]
AFAIK tensors are only supported in Paraview via plugins. Both of the plugins listed [here](https://www.paraview.org/Wiki/ParaView/User_Created_Plugins) have some example VTK files that can be visualised. So thats maybe not more useful than the examples that you already found, but maybe it hints to the fact that the tensor reader in Paraview would be VTK only?

**Comment by jppelteret** [comment]
*fact -> idea

**Comment by davydden** [comment]
@jppelteret looks like `visit` has something, not sure it's what we need, but there is `Add->Tensor-><name>`. I really hope one does not need to fiddle with a custom build of paraview to visualize tensors, that is major PITA...

**Comment by jppelteret** [comment]
> I really hope one does not need to fiddle with a custom build of paraview to visualize tensors

Hmm... I've just loaded up Paraview 5.5, and it does have a `tensor glyph` filter built in. That's a nice surprise, because the last time I visualised tensors I did have to hand compile Paraview with the plugins enabled.

So, back to your original question... Sorry, I have no idea as to what this should look like in the VTU format :-(

**Comment by davydden** [comment]
> it does have a tensor glyph filter built in.

ok, so there is a hope 👍

**Comment by davydden** [comment]
@jppelteret 

>  last time I visualised tensors

do you happen to have any 2d / 3d tensor example you visualized I can use for tests?

**Comment by jppelteret** [comment]
> ok, so there is a hope 👍

Just tested it... it works! (for VTK files :-P )

screen shot 2018-06-21 at 15 19 08

**Comment by jppelteret** [comment]
> do you happen to have any 2d / 3d tensor example you visualized I can use for tests?

Check out the link to the plugins I mentioned in this [comment](https://github.com/dealii/dealii/pull/6807#issuecomment-399088739). There are some VTK examples there.

**Comment by davydden** [comment]
@jppelteret works now: Left - output from the test, right - output from the input file from one of plugins.

screen shot 2018-06-21 at 16 50 44

I still need to figure out how to actually make a blessed output as `vtu` is compressed. @bangerth any hints?

**Comment by bangerth** [comment]
I think the example in the documentation of the `DataPostprocessorTensor` class discusses how to do this in visit. You had to hand-combine scalar fields into a vector.

**Comment by davydden** [comment]
p.s. i should still fix a warning
```
In file included from /Users/davydden/libs-sources/deal.ii/deal.ii/source/base/data_out_base.cc:34:
In file included from /Users/davydden/libs-sources/deal.ii/deal.ii/include/deal.II/base/data_out_base.h:22:
In file included from /Users/davydden/libs-sources/deal.ii/deal.ii/include/deal.II/base/geometry_info.h:23:
In file included from /Users/davydden/libs-sources/deal.ii/deal.ii/include/deal.II/base/point.h:23:
/Users/davydden/libs-sources/deal.ii/deal.ii/include/deal.II/base/tensor.h:1374:31: warning: remainder by zero is undefined [-Wdivision-by-zero]
      indices[r] = (remainder % dim);
                              ^ ~~~
/Users/davydden/libs-sources/deal.ii/deal.ii/source/base/data_out_base.cc:5359:43: note: in instantiation of member function 'dealii::Tensor<2, 0, double>::unrolled_to_component_indices' requested here
                          Tensor<2, dim>::unrolled_to_component_indices(c)[0];
                                          ^
/Users/davydden/libs-sources/deal.ii/deal.ii/source/base/data_out_base.cc:6905:18: note: in instantiation of function template specialization 'dealii::DataOutBase::write_vtu_main<0, 1>' requested here
    DataOutBase::write_vtu_main(get_patches(),
                 ^
```

File: source/base/data_out_base.cc [hunk 5]
```diff
@@ -5597,43 +5622,80 @@ namespace DataOutBase
             out << data_names[last_component];
           }
 
-        out << "\" NumberOfComponents=\"3\" format=\"" << ascii_or_binary
-            << "\">\n";
+        out << "\" NumberOfComponents=\"" << (is_tensor ? "9" : "3")
```
Comment (jppelteret) [inline source/base/data_out_base.cc]: Please create a variable `n_components` and store the appropriate value in there.

File: source/base/data_out_base.cc [hunk 6]
```diff
@@ -5597,43 +5622,80 @@ namespace DataOutBase
...
+            << "\" format=\"" << ascii_or_binary << "\">\n";
 
         // now write data. pad all vectors to have three components
         std::vector<float> data;
-        data.reserve(n_nodes * dim);
+        data.reserve(n_nodes * (is_tensor ? 9 : 3));
```
Comment (jppelteret) [inline source/base/data_out_base.cc]: Please use the aforementioned `n_components` here.

File: source/base/data_out_base.cc [hunk 7]
```diff
@@ -5571,11 +5593,14 @@ namespace DataOutBase
...
         const auto name            = std::get<2>(range);
+        const bool is_tensor =
+          (std::get<3>(range) ==
+           DataComponentInterpretation::component_is_part_of_tensor);
         AssertThrow(last_component >= first_component,
                     ExcLowerRange(last_component, first_component));
         AssertThrow(last_component < n_data_sets,
                     ExcIndexRange(last_component, 0, n_data_sets));
-        AssertThrow(last_component + 1 - first_component <= 3,
+        AssertThrow((last_component + 1 - first_component <= 3 || is_tensor),
```
Comment (jppelteret) [inline source/base/data_out_base.cc]: Please do two things here:
1. conditionally check the assert using `if (!is_tensor)`
2. make the equivalent check for `if (is_tensor)`, as this should have no more than 9 components

File: source/base/data_out_base.cc [hunk 8]
```diff
@@ -5571,11 +5593,14 @@ namespace DataOutBase
         const auto first_component = std::get<0>(range);
         const auto last_component  = std::get<1>(range);
         const auto name            = std::get<2>(range);
+        const bool is_tensor =
+          (std::get<3>(range) ==
+           DataComponentInterpretation::component_is_part_of_tensor);
```
Comment (jppelteret) [inline source/base/data_out_base.cc]: I guess there should be an additional assertion here, that if `!is_tensor` then `std::get<3>(range) == DataComponentInterpretation::component_is_part_of_vector`

File: tests/data_out/data_out_12.cc [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  data_out.write_vtu(deallog.get_file_stream());
+}
+
+
+
+int
+main()
+{
+  initlog();
+  check<3>();
```
Comment (jppelteret) [inline tests/data_out/data_out_12.cc]: Have you also checked that the 2d output is readable by Paraview? I think that its probably a good idea to verify this.

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 10]
```diff
@@ -1212,21 +1212,27 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::
...
-               "be visualized by their real and imaginary parts separately."));
+      // as vector or tensor fields cannot be complex-valued, because we cannot
+      // visualize complex-valued vector fields
+      Assert(
+        !(((std::find(
+              data_component_interpretation.begin(),
+              data_component_interpretation.end(),
+              DataComponentInterpretation::component_is_part_of_vector) !=
+            data_component_interpretation.end()) ||
+           (std::find(
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/numerics/data_out_dof_data.templates.h]: In particular, your condition is `! (A || B)`, which is the same as `!A && !B`, and because you're now enforcing two conditions, they can be checked in separate assertions.

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 11]
```diff
@@ -1391,6 +1397,50 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::
           output_component += patch_space_dim;
           i += patch_space_dim;
         }
+      else if ((*d)->data_component_interpretation[i] ==
+               DataComponentInterpretation::component_is_part_of_tensor)
+        {
+          const unsigned int size = patch_space_dim * patch_space_dim;
+          // ensure that there is a continuous number of next
+          // space_dim*space_dim components that all deal with vectors
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h]: -> tensors

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 12]
```diff
@@ -1391,6 +1397,50 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::
...
+          Assert(i + size <= (*d)->n_output_variables,
+                 Exceptions::DataOutImplementation::ExcInvalidVectorDeclaration(
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h]: here and a few lines down, you'll need a separate exception class

File: source/base/data_out_base.cc [hunk 13]
```diff
@@ -5302,6 +5304,26 @@ namespace DataOutBase
 
 
 
+  namespace internal
+  {
+    template <int dim>
+    TableIndices<2>
+    unrolled_to_component_indices(const unsigned int i)
```
Comment (bangerth) [inline source/base/data_out_base.cc]: That's a bad name because it's not clear whether you want to transfer between vector, rank-2 tensor, etc objects and plain index space. Please use the corresponding functions directly from the tensor classes.

File: source/base/data_out_base.cc [hunk 14]
```diff
@@ -5302,6 +5304,26 @@ namespace DataOutBase
...
+    {
+      return Tensor<2, dim>::unrolled_to_component_indices(i);
+    }
+
+    template <>
+    TableIndices<2>
+    unrolled_to_component_indices<0>(const unsigned int /*i*/)
```
Comment (bangerth) [inline source/base/data_out_base.cc]: same here, of course

File: source/base/data_out_base.cc [hunk 15]
```diff
@@ -5597,43 +5622,80 @@ namespace DataOutBase
...
+                    Assert(false, ExcInternalError());
+                  }
 
-                default:
-                  // VTK doesn't support anything else than vectors with 1, 2,
-                  // or 3 components
-                  Assert(false, ExcInternalError());
+                // now put the tensor into data
+                for (unsigned int i = 0; i < 3; ++i)
+                  for (unsigned int j = 0; j < 3; ++j)
```
Comment (bangerth) [inline source/base/data_out_base.cc]: Can you add a comment here saying that you are padding with zeros because VTK format *always* wants to see a 3x3 tensor, regardless of dimension?

File: source/base/data_out_base.cc (hunk 13 above)
Comment (davydden) [reply source/base/data_out_base.cc]: > Please use the corresponding functions directly from the tensor classes.

I explicitly do NOT use it directly as Tensor2 does not like `dim=2` and there is a compiler warning of division by zero. Otherwise I would never introduce this wrapper in the first place.

File: source/base/data_out_base.cc (hunk 14 above)
Comment (davydden) [reply source/base/data_out_base.cc]: ditto

File: source/base/data_out_base.cc (hunk 13 above)
Comment (bangerth) [reply source/base/data_out_base.cc]: I assumed you had good reasons to do things this way, but let's try to find ways to make this work somehow. What is the error message you are getting?

Comment (davydden) [reply source/base/data_out_base.cc]:
```
      indices[r] = (remainder % dim);
      remainder /= dim;
```
in Tensor are not nice for `dim==0`. DataOut is instantiated for dim==0.

Comment (bangerth) [reply source/base/data_out_base.cc]: Could we specialize this in the tensor class?

If you can't figure this out easily, I'd be ok if you renamed the current function to `rank_2_unrolled_...` and opened a github issue that reminds us that this could be cleaned up later on.

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 16]
```diff
@@ -1391,6 +1397,50 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::
...
+          std::string name = (*d)->names[i];
+          for (unsigned int dd = 1; dd < size; ++dd)
+            if (name != (*d)->names[i + dd])
+              {
+                name = "";
+                break;
+              }
+
+          // finally add a corresponding range
+          std::tuple<unsigned int,
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h]: You can simplify the next 11 lines a bit using the same ideas as in #6876.

File: source/base/data_out_base.cc (hunk 8 above)
Comment (davydden) [reply source/base/data_out_base.cc]: >  there should be an additional assertion here

why? `is_tensor` just a short-cut for range being a tensor. Eventually somebody will add another interpretation so I don't see the reason of having this assert.

**Comment by davydden** [comment]
@jppelteret @bangerth I updated the PR, please give it another look.

@jppelteret here's a 2D output (added to the test):

screen shot 2018-07-04 at 10 07 13

File: include/deal.II/base/tensor.h [hunk 17]
```diff
@@ -1357,6 +1357,45 @@ Tensor<rank_, dim, Number>::component_to_unrolled_index(
...
+namespace internal
+{
+  // unrolled_to_component_indices is instantiated from DataOut for dim==0
+  // and rank=2. Make sure we don't have compiler warnings.
+
+  template <int dim>
+  inline unsigned int
+  mod(const unsigned int &remainder)
+  {
+    return remainder % dim;
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: `remainder` is an odd choice for the variable name here because the variable is the *operand*, whereas the remainder is the *result* of the operation. Just call it `x` or something. Also, no need for the reference.

File: include/deal.II/base/tensor.h [hunk 18]
```diff
@@ -1357,6 +1357,45 @@ Tensor<rank_, dim, Number>::component_to_unrolled_index(
...
+  {
+    Assert(false, ExcInternalError());
+    return remainder;
+  }
+
+  template <int dim>
+  inline void
+  div(unsigned int &remainder)
+  {
+    remainder /= dim;
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Just write this as
```
  inline unsigned int
  div(const unsigned int x)
  {
    return x / dim;
```

File: include/deal.II/base/tensor.h [hunk 19]
```diff
@@ -1357,6 +1357,45 @@ Tensor<rank_, dim, Number>::component_to_unrolled_index(
...
+  }
+
+  template <>
+  inline void
+  div<0>(unsigned int & /*remainder*/)
+  {
+    Assert(false, ExcInternalError());
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: ...with the obvious adjustment here...

File: include/deal.II/base/tensor.h [hunk 20]
```diff
@@ -1369,8 +1408,8 @@ Tensor<rank_, dim, Number>::unrolled_to_component_indices(const unsigned int i)
   unsigned int remainder = i;
   for (int r = rank_ - 1; r >= 0; --r)
     {
-      indices[r] = (remainder % dim);
-      remainder /= dim;
+      indices[r] = internal::mod<dim>(remainder);
+      internal::div<dim>(remainder);
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: and then do
```
  remainder = internal::div<dim>(remainder);
```
here. I think that would be easier to read. It would also restore the symmetry between the `mod` and `div` operations in that both of them return things by value.

**Comment by davydden** [comment]
@bangerth done.

File: include/deal.II/numerics/data_out_dof_data.h [hunk 21]
```diff
@@ -149,6 +149,19 @@ namespace Exceptions
...
                    << "> does not satisfy these conditions.");
+
+    DeclException2(ExcInvalidTensorDeclaration,
+                   int,
+                   std::string,
+                   << "When declaring that a number of components in a data "
+                   << "set to be output logically form a tensor instead of "
+                   << "simply a set of scalar fields, you need to specify "
+                   << "this for all relevant components. Furthermore, "
+                   << "vectors must always consist of exactly <dim*dim> "
```
Comment (jppelteret) [inline include/deal.II/numerics/data_out_dof_data.h]: Am I correct that here you mean to say
> tensors must always

?

File: include/deal.II/numerics/data_out_dof_data.h [hunk 22]
```diff
@@ -149,6 +149,19 @@ namespace Exceptions
...
+                   << "components. However, the vector component at "
```
Comment (jppelteret) [inline include/deal.II/numerics/data_out_dof_data.h]: the vector component -> the tensor component (?)

**Comment by jppelteret** [comment]
Thanks for making the suggested changes, and for adding the 2d test!

**Comment by davydden** [comment]
@jppelteret done.
